src/node/desktop/src/main/toolbar-manager.ts

```diff
@@ -0,0 +1,160 @@
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+/* eslint-disable max-len */
+
+import { BrowserWindow } from 'electron';
+import * as fs from 'fs';
+
+export interface ToolbarData {
+  buttons: {
+    iconPath: string;
+    tooltip: string;
+    onClick: string;
+  }[];
+}
+
+export class ToolbarManager {
+  constructor() {}
+
+  async createToolbar(window: BrowserWindow, toolbarData: ToolbarData) {
+    const jsScript =
+      'console.log("windowOnload");' +
+      this.addToolbarJsAsText() +
+      ';' +
+      this.addStylesJsAsText() +
+      ';' +
+      this.addButtonsJsAsText(toolbarData);
+
+    try {
+      await window.webContents.executeJavaScript(jsScript);
+
+      // eslint-disable-next-line @typescript-eslint/no-implicit-any-catch
+    } catch (err: any) {
+      if (err.message !== 'An object could not be cloned.') {
+        console.log('Error message: ', err.message);
+        console.error('Error ', err, ' when trying to create toolbar with data: ', toolbarData);
+        console.log('---------------------------');
+        console.log('JS Script for Toolbar: ');
+        console.log(jsScript);
+      }
+    }
+  }
+
+  addToolbarJsAsText() {
+    const toolbar = ` 
+      const toolbar = document.createElement('div');
+      toolbar.id = 'custom-toolbar';
+      document.body.prepend(toolbar);
+    `;
+
+    return toolbar;
+  }
+
+  addStylesJsAsText() {
+    const styles = `
+        body{
+          padding-top:29px;
+        }
+
+        #custom-toolbar {
+          display: flex;
+          align-items: center;
+          justify-content: flex-start;
+          width: 100%;
+          background: gray;
+          padding: 4px;
+          background-color: rgb(239, 239, 239);
+          border-bottom: 1px solid rgb(214, 214, 214);
+          position: fixed;
+          top:0;
+          left:0;
+        }
+
+        .custom-toolbar_button {
+          display: flex;
+          align-items: center;
+          justify-content: center;
+
+          cursor: pointer;
+          margin-right: 8px;
+
+          -webkit-user-select: none;
+          -moz-user-select: none;
+          -ms-user-select: none;
+          user-select: none;
+        }
+
+        .custom-toolbar_icon {
+          width: 26px;
+          height: 20px;
+        }
+
+        .custom-toolbar_icon:active {
+          filter: brightness(80%) saturate(150%);
+        }
+
+        @media (prefers-color-scheme: dark) {
+          #custom-toolbar {
+            background-color: rgb(48, 48, 48);
+            border-bottom: 1px solid rgb(12, 31, 48);
+          }
+        }
+    `;
+
+    return `
+      const styleSheet = document.createElement('style');
+      styleSheet.innerText = \`${styles}\`;
+      document.head.appendChild(styleSheet);
+      `;
+  }
+
+  addButtonsJsAsText(toolbarData: ToolbarData) {
+    let addButtonsFn = `
+      const createButton = (iconPath, tooltip, onClick) => {
```

Comment:
The buttons need to be annotated so they work correctly with screen readers, and be usable via the keyboard.

Strongly recommend using semantic html and making the outer "div" be a `<button>`, instead. That should give most of the necessary behavior automatically.

* each button needs to be reachable via the Tab key (tabindex="0")
* when given focus via the keyboard, a focus rectangle must be displayed (we use the `focus-visible` polyfill to accomplish this; we don't want focus rectangle shown if you just click on the button)
* when a button has focus you can activate it via keyboard (spacebar and enter key)
* each button must have descriptive text for screen readers (the alt text on the image might be sufficient, I would need to test once the rest of the work is done)

Reply:
More on buttons:

* https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role
* https://www.w3.org/TR/wai-aria-practices/#button

Reply:
Done in https://github.com/rstudio/rstudio/pull/10541/commits/6deede4405d4f6cfc08f5c0d2759f7ae4a7d7c6e